src/my-bot.ts

```diff
@@ -1,38 +1,423 @@
+        apiKey: "AIzaSyD6ACQdu7gK-BgtJs-3Hu1Lkczk8fp0Abo",
+        authDomain: "brokenbot-battleships.firebaseapp.com",
+        databaseURL: "https://brokenbot-battleships.firebaseio.com",
+        storageBucket: "brokenbot-battleships.appspot.com"
+    };
+    private matchId: number;
+
+    constructor() {
+        firebase.initializeApp(this.config);
+        this.authenticate();        
+    }
+    
+    public getShipPositions(): ShipPlace[] {
+        let exists: boolean = false;
+        let counter: number = 0;        
+        do {
+            this.matchId = Math.floor(Math.random() * 10000) + 1;
+            firebase.database().ref('matches/' + this.matchId.toString()).once('value').then((snapshot) => {
+                if (snapshot.val()) {
+                    exists = true;
+                } else {
+                    exists = false;
+                }
+            });
+            counter++;
+            if (counter > 10000) {
+                throw new Error("Infinite loop when constructing bot");
+            }
+        } while (exists)
+        firebase.database().ref('matches/' + this.matchId.toString()).set({
+            started: true,
+            hitmode: false,
+            sizes: {
+                patrol: true,
+                submarine: true,
+                destroyer: true,
+                battleship: true,
+                carrier: true
+            },
+            hitmap: []
+        });
+        let shipPlaces: ShipPlace[] = [];
+        let done: boolean = false;
+        counter = 0;
+        do {
+            shipPlaces = [];
+            counter++;
+            shipPlaces[0] = this.getShipPlace(shipPlaces,5);
+            if (shipPlaces[0].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[1] = this.getShipPlace(shipPlaces,4);
+            if (shipPlaces[1].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[2] = this.getShipPlace(shipPlaces,3);
+            if (shipPlaces[2].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[3] = this.getShipPlace(shipPlaces,3);
+            if (shipPlaces[3].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[4] = this.getShipPlace(shipPlaces,2);
+            if (shipPlaces[4].StartingSquare.Column === -1) {
+                continue;
+            }
+            done = true;
+        } while ((!done) && (counter < 10000))
+        if (counter === 10000) {
+            throw new Error("Infinite loop generating ship positions");
         }
-        return { Row: "A", Column: 1 };  
+        return shipPlaces;
     }
 
-    private getNextTarget(position) {
+    public selectTarget(gamestate): Promise<Position> {
+        return new Promise((resolve, reject) => {
+            firebase.database().ref('matches/' + this.matchId.toString()).once('value').then((snapshot) => { 

```

Comment:
I am impressed at the use of firebase - you've really engaged with it in some quite cool ways.